File: 19_Bunny/csharp/Bunny.cs
namespace Bunny
{
    internal class Bunny
    {
        private const int asciiBase = 64;
        private readonly int[] bunnyData = {
            2,21,14,14,25,
            1,2,-1,0,2,45,50,-1,0,5,43,52,-1,0,7,41,52,-1,
            1,9,37,50,-1,2,11,36,50,-1,3,13,34,49,-1,4,14,32,48,-1,
            5,15,31,47,-1,6,16,30,45,-1,7,17,29,44,-1,8,19,28,43,-1,
            9,20,27,41,-1,10,21,26,40,-1,11,22,25,38,-1,12,22,24,36,-1,
            13,34,-1,14,33,-1,15,31,-1,17,29,-1,18,27,-1,
            19,26,-1,16,28,-1,13,30,-1,11,31,-1,10,32,-1,
            8,33,-1,7,34,-1,6,13,16,34,-1,5,12,16,35,-1,
            4,12,16,35,-1,3,12,15,35,-1,2,35,-1,1,35,-1,
            2,34,-1,3,34,-1,4,33,-1,6,33,-1,10,32,34,34,-1,
            14,17,19,25,28,31,35,35,-1,15,19,23,30,36,36,-1,
            14,18,21,21,24,30,37,37,-1,13,18,23,29,33,38,-1,
            12,29,31,33,-1,11,13,17,17,19,19,22,22,24,31,-1,
            10,11,17,18,22,22,24,24,29,29,-1,
            22,23,26,29,-1,27,29,-1,28,29,-1,4096
        };

        public void Run()
        {
            PrintString(33, "BUNNY");
            PrintString(15, "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
            PrintLines(3);

            // Set up a BASIC-ish data object
            BasicData data = new (bunnyData);

            // Get the first five data values into an array.
            // These are the characters we are going to print.
            // Unlike the original program, we are only converting
            // them to ASCII once.
            var a = new char[5];
            for (var i = 0; i < 5; ++i)
            {
                a[i] = (char)(asciiBase + data.Read());
            }
            PrintLines(6);

            PrintLines(1);
            var col = 0;
            while (true)
            {
                var x = data.Read();
                if (x < 0) // Start a new line
                {
                    PrintLines(1);
                    col = 0;
                    continue;
                }
                if (x > 128) break; // End processing
                col += PrintSpaces(x - col); // Move to TAB position x (sort of)
                var y = data.Read(); // Read the next value
                for (var i = x; i <= y; ++i)
                {
                    // var j = i - 5 * (i / 5); // BASIC didn't have a modulus operator
                    Console.Write(a[i % 5]);
                    // Console.Write(a[col % 5]); // This works, too
                    ++col;
                }
            }
            PrintLines(6);
        }
        private static void PrintLines(int count)
        {
            for (var i = 0; i < count; ++i)
                Console.WriteLine();
        }
        private static int PrintSpaces(int count)
        {
            for (var i = 0; i < count; ++i)
                Console.Write(' ');
            return count;
        }
        public static void PrintString(int tab, string value, bool newLine = true)
        {
            PrintSpaces(tab);
            Console.Write(value);
            if (newLine) Console.WriteLine();
        }

    }
}

File: 19_Bunny/python/bunny.py
#!/usr/bin/env python3


# This data is meant to be read-only, so we are storing it in a tuple
import json

with open("data.json") as f:
    DATA = tuple(json.load(f))


def print_intro() -> None:
    print(" " * 33 + "BUNNY")
    print(" " * 15 + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY")
    print("\n\n")


def main() -> None:
    print_intro()

    # Using an iterator will give us a similar interface to BASIC's READ
    # command. Instead of READ, we will call 'next(data)' to fetch the next element.
    data = iter(DATA)

    # Read the first 5 numbers. These correspond to letters of the alphabet.
    # B=2, U=21, N=14, N=14, Y=25

    # Usually, list comprehensions are good for transforming each element in a sequence.
    # In this case, we are using range to repeat the call to next(data) 5 times. The underscore (_)
    # indicates that the values from range are discarded.
    bunny = [next(data) for _ in range(5)]
    L = 64

    # Interpretting a stream of data is a very common software task. We've already intepretted
    # the first 5 numbers as letters of the alphabet (with A being 1). Now, we are going to
    # combine this with a different interpretation of the following data to draw on the screen.
    # The drawing data is essentially a series of horizontal line segments given as begin and end
    # offsets.
    while True:
        command = next(data)

        if command < 0:
            print()
            continue

        if command > 128:
            break

        # If we've reached this portion of the code, 'command' indicates the 'start'
        # position of a line segment.
        start = command
        # Position cursor at start
        print(" " * start, end="")

        # The following number, indicates the end of the segment.
        end = next(data)
        # Unlike FOR I=X TO Y, the 'stop' argument of 'range' is non-inclusive, so we must add 1
        for i in range(start, end + 1, 1):
            # Cycle through the letters in "BUNNY" as we draw line
            j = i - 5 * int(i / 5)
            print(chr(L + bunny[j]), end="")


if __name__ == "__main__":
    main()
